### PR TITLE
Use crosstool-ng to build latest m68k version of gcc for repo workflow

### PR DESCRIPTION
**Pull request by iratahack**
These changes build the latest available (gcc 13.2.0) m68k dev tools replacing the 6.x.x version in the docker container environment. 
Also uses cmake to build the host based executables and SGDK libraries.

@joeyparrish , please take a look at the changes and let me know if you see anything obvious.
@Stephane-D , I think I have the build of xgm2tool.jar working let me know if you have any comments.

Is a C-library required? Newlib?

### CODE REVIEW

File: .github/workflows/sgdk-docker.yml [hunk 1]
```diff
@@ -36,6 +36,8 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+          submodules: 'true'
```
Comment (iratahack) [inline .github/workflows/sgdk-docker.yml:40]: Checkout submodules for Sjasm and musicg

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,6 @@
 [submodule "tools/sjasm"]
 	path = tools/sjasm
 	url = https://github.com/Konamiman/Sjasm.git
+[submodule "tools/musicg"]
```
Comment (iratahack) [inline .gitmodules:4]: Submodule required to build xgm2tool.jar

File: Dockerfile [hunk 3]
```diff
@@ -25,29 +25,14 @@ RUN jlink \
 
 # Stage One - build all tools and libs for SGDK
 FROM jre-minimal as build
-RUN apk add --no-cache build-base git
+RUN apk add --no-cache build-base git cmake ninja
```
Comment (iratahack) [inline Dockerfile:28]: Use cmake and ninja for building the libs.

File: Dockerfile [hunk 4]
```diff
@@ -25,29 +25,14 @@ RUN jlink \
...
-COPY . /sgdk
-
-# Building bintos
-WORKDIR $SGDK_PATH/tools/bintos
-RUN gcc -O2 -s src/bintos.c -o $SGDK_PATH/bin/bintos
-
-# Building xgmtool
-WORKDIR $SGDK_PATH/tools/xgmtool
-RUN gcc -fexpensive-optimizations -Os -s src/*.c -o $SGDK_PATH/bin/xgmtool
+COPY . $SGDK_PATH
```
Comment (iratahack) [inline Dockerfile:35]: Use env var instead of hard coded path.

File: Dockerfile [hunk 5]
```diff
@@ -72,19 +57,29 @@ RUN find . -name "*.java" | xargs javac
...
-RUN mkdir lib
-#build libmd.a
-RUN make -f makelib.gen release
-#build libmd_debug.a
-RUN make -f makelib.gen debug
-RUN rm -rf $SGDK_PATH/tools
+# 32-bit not supported on Alpine, this may have side-effects
+RUN sed -i /-m32/d $SGDK_PATH/tools/xgmtool/src/CMakeLists.txt
+ENV PREFIX=m68k-elf-
+RUN cmake -G Ninja -S $SGDK_PATH -B $SGDK_PATH/build_release -DCMAKE_BUILD_TYPE=Release && cmake --build $SGDK_PATH/build_release
```
Comment (iratahack) [inline Dockerfile:77]: Build the libs here.

File: deps/gcc.Dockerfile [hunk 6]
```diff
@@ -1,151 +1,39 @@
...
-RUN make DESTDIR="$INSTALLDIR/$GCC-bootstrap" install-strip-gcc
-RUN rm -rf "$INSTALLDIR/$GCC-bootstrap/usr/share"
-RUN cp -r "$INSTALLDIR/$GCC-bootstrap/usr/"* /usr/
 #################################################################################
-#############################    Building NewLib    #############################
+############################    Building crosstool-ng    ########################
 #################################################################################
-WORKDIR $BUILDDIR/$NEWLIB
-ENV CFLAGS_FOR_TARGET="-Os -g -ffunction-sections -fdata-sections -fomit-frame-pointer -ffast-math"
+WORKDIR /root
```
Comment (iratahack) [inline deps/gcc.Dockerfile:17]: Build crosstool-ng. Not currently supported as a package in this version of Alpine.

File: deps/gcc.Dockerfile [hunk 7]
```diff
@@ -1,151 +1,39 @@
...
-RUN rm "$INSTALLDIR/$GCC/usr/lib/libcc1.so.0.0.0"
-
-RUN cp -ar "$INSTALLDIR/$GCC/usr/" "$STAGINGDIR"
-
+WORKDIR $BASEDIR
+ENV CT_EXPERIMENTAL=y
+ENV CT_ALLOW_BUILD_AS_ROOT=y
+ENV CT_ALLOW_BUILD_AS_ROOT_SURE=y
+# Configure and build the tools
+RUN ct-ng m68k-elf
```
Comment (iratahack) [inline deps/gcc.Dockerfile:33]: Configure and build the tools.

Comment (Stephane-D) [reply deps/gcc.Dockerfile:33]: not sure what GCC 13.2 brings for m68j compared to GCC 6.3, from my tests later version perform worse on code generation but I admit I only tested up to GCC 10.
Would be nice to compare 'benchmark' sample compiled with GCC 6.3 versus GCC 13.2 to see the difference.
My way of seeing things is always : the simpler and the lighter, the better...
I'm almost certain the GCC 13.2 toolchain is heavier than GCC 6.x but if that comes with some benefits why not.
Also i'm not sure about CMake requirement, again, what does it brings here ?

Comment (iratahack) [reply deps/gcc.Dockerfile:33]: @Stephane-D , In general I tend to agree. I did run the benchmark sample and there are some improvements and a few things are worse. GCC 13.2.0 gives a slightly better overall score of 15247 compared to 14298 from GCC 6.3. These results are from blastem on Windows. I need to try on real hardware.

I think of CMake as a build description language. You describe how to build your project and CMake handles the generation of the actual build scripts, including handling all the dependancies. It supports generating build scripts for a number of standard tools like Make and Ninja. I find Ninja to be faster than using Make in many cases, especially in msys or mingw environments. It's also good at keeping the derived files contained so you can do a debug or a release build without having to perform a clean inbetween, etc.

Let me know what you think. I've you don't want to move to gcc 13.2 I can change the config to build 6.5 which is closer to what you have now.

Comment (Stephane-D) [reply deps/gcc.Dockerfile:33]: The better performance on GCC 13.2 is a good surprise ! So definitely you can keep it.
About CMake I understand it interest if it helps for the Docker usage, so it's all good for me, thanks :)

**Review by Stephane-D — APPROVED** [review]

**Comment by Stephane-D** [comment]
I removed git sub modules, the make the repo more complex to deal and we don't need to update both musicg / sjasm.
I keep musicg in sources format directly into XGM2Tool folder. Sjasm is also now in sources format in _tools_ folder.
Note that the sgdk-docker git task is failing currently.

**Comment by Stephane-D** [comment]
> These changes build the latest available (gcc 13.2.0) m68k dev tools replacing the 6.x.x version in the docker container environment. Also uses cmake to build the host based executables and SGDK libraries.
> 
> @joeyparrish , please take a look at the changes and let me know if you see anything obvious. @Stephane-D , I think I have the build of xgm2tool.jar working let me know if you have any comments.
> 
> Is a C-library required? Newlib?

Forgot to reply here but only libgcc is needed and only a very sub part of it is really used by SGDK.

**Comment by iratahack** [comment]
> I removed git sub modules, the make the repo more complex to deal and we don't need to update both musicg / sjasm.
> 
> I keep musicg in sources format directly into XGM2Tool folder. Sjasm is also now in sources format in _tools_ folder.
> 
> Note that the sgdk-docker git task is failing currently.

I'll take a look.

**Comment by Stephane-D** [comment]
If it helps :
image

maybe a typo in the repo name ?

**Pull request merged**